pages/exercises/[lessonSlug].tsx

```diff
@@ -70,31 +71,34 @@ const Exercises: React.FC<QueryDataProps<GetExercisesQuery>> = ({
       explanation: exercise.explanation || '',
       userAnswer: userAnswers[exercise.id] ?? null
     }))
-
-  const exercise = currentExercises[exerciseIndex]
+    .filter(
+      exercise => !hideAnswered || exercise.userAnswer !== exercise.answer
+    )
 
   return (
     <Layout title={currentLesson.title}>
-      {exercise ? (
+      {solvingExercise ? (
         <Exercise
-          key={exerciseIndex}
-          exercise={exercise}
-          setExerciseIndex={setExerciseIndex}
           lessonTitle={currentLesson.title}
-          hasPrevious={exerciseIndex > 0}
-          hasNext={exerciseIndex < currentExercises.length - 1}
-          submitUserAnswer={(userAnswer: string) => {
-            setUserAnswers({ ...userAnswers, [exercise.id]: userAnswer })
+          exercises={currentExercises}
+          userAnswers={userAnswers}
+          onExit={localUserAnswers => {
+            setUserAnswers({ ...userAnswers, ...localUserAnswers })
+            setSolvingExercise(false)
+          }}
```

Comment:
`onExit` gets called when we exit the `Exercise` component and go back to the `ExerciseList` view mode. The `userAnswers` gets updated only after we exit the `Exercise` component because if we don't want `userAnswers` changing which would change `currentExercises` which could cause the exercise that gets shown in the `Exercise` component to change in a way we don't want.